DM-25776: Refactor DataCoordinates and add custom containers for them.

TallJimbo opened the pull request.

TallJimbo committed:
Add custom containers for DataCoordinates.

These provide some additional APIs and consistency guarantees relative
to native containers, setting the stage for specialized
implementations for query results later.

TallJimbo committed:
Add method to obtain just dimension values from a query row.

TallJimbo committed:
Handle edge case where we can use DataCoordinate.subset.

This avoids downgrading an ExpandedDataCoordinate into a regular
DataCoordinate in this case as well.

TallJimbo committed:
Fix visit_definition edge case in DimensionRecord initialization.

Before we added visit_definition, it was safe to assume that taking
a DimensionElement's required dimensions, expanding them, and then
removing any dimension implied by another would yield the element's
required dimensions.  That's no longer always true, and that's
revealed a lot of fragile code that depended on the ordering of
dependent dimensions in dimension records.  Most of that was fixed
on DM-15536, but this one escaped.

TallJimbo committed:
Remove DataCoordinate.fingerprint and DatasetRef.hash.

These are no longer in use after the dataset table restructuring.

TallJimbo committed:
Add default implementation for NamedKeyMapping.byName().

TallJimbo committed:
Add fields attribute and stringification to DimensionRecord.

TallJimbo committed:
Add copy method to SimpleQuery.

copy.copy doesn't go deep enough to be useful; copy.deepcopy almost
certainly goes much deeper than it needs to.

TallJimbo committed:
Various minor fixes to types and comments.

TallJimbo committed:
Make SimpleQuery's Select tag class a nested class.

This makes usage a bit more verbose, but it avoids confusion with
sqlalchemy.sql.Select.

TallJimbo committed:
Move SimpleQuery from registry subpackage to core.

This paves the way for core.dimensions to depend on it.

TallJimbo committed:
Vectorize fetching of DimensionRecords.

This does not help any higher-level APIs yet, but it sets the stage
for much faster data ID expansion of query results when those are
represented by DataCoordinateIterable classes in the future.

TallJimbo committed:
Handle missing implied dimensions more carefully in expandDataId.

Prior to the DataCoordinate refactor the case where two steps of
implied dimensions were both null was apparently handled accidentally,
and this explicit handling of that case is an improvement (that also
fixes tests downstream in the verify package).

TallJimbo committed:
Remove unnecessary data ID expansion in butler lookups.

We need to expand data IDs (at least in general) when writing datasets,
so we can fill out file templates.  But all those database lookups are
totally unnecessary in get and datasetExists.

TallJimbo committed:
Add realistically complex dimension data dump for tests.

TallJimbo committed:
Refactor DataCoordinate into a single interface.

...with multiple implementations.  See the long comment at the top of
coordinate.py for design rationale.

The old versions of these classes had no direct unit test coverage,
but a ton of indirect usage in other tests.  I'll be adding more
direct tests later on this ticket, once additional functionality that
will make them easier to write lands.

TallJimbo committed:
Add unit tests for DataCoordinate.

TallJimbo committed:
Add tests for DataCoordinate containers.

These aren't explicitly complete, but these classes are used
extensively in the tests for DataCoordinates, so everything that isn't
covered is at most trivial one-line delegating method.

timj commented on doc/lsst.daf.butler/dimensions.rst:
```diff
@@ -55,14 +55,13 @@ Different datasets with the same `DatasetType` are always identified by the same
-`ExpandedDataCoordinate` also has additional attributes that hold all of the metadata fields associated with its dimensions.
+The data IDs returned by the `Butler` or `Registry` (and most of those used internally) are usually instances of the `DataCoordinate` class.
+`DataCoordinate` instances can have different states of knowledge about the dimensions they identify.
+They always contains at least the key-value pairs that correspond to its `DimensionGraph`'s `~DimensionGraph.required` subset -- that is, the minimal set of keys needed to fully identify all other dimensions in the graph.
```
"contains" -> "contain".

Also, doesn't there need to be a `\ ` or something between `DimensionGraph` and the `s`?

TallJimbo replied:
It builds fine for me without the `\ ` in all of my local `package-docs` builds, even though I know it's not supposed to.  I suppose I should add it anyway.  Don't know why I've never encountered that problem.

timj replied:
I bet it's the non-alphabetic apostrophe character that is the difference.

timj commented on python/lsst/daf/butler/core/dimensions/_dataCoordinateIterable.py:
```diff
@@ -0,0 +1,764 @@
+    def __init__(self, dataIds: Collection[DataCoordinate], graph: DimensionGraph, *,
+                 hasFull: Optional[bool] = None, hasRecords: Optional[bool] = None,
+                 check: bool = True):
+        self._dataIds = dataIds
```
Is it at all possible for `dataIds` to be a mutable collection which can then be modified by the caller? ie should this be `self._dataIds = tuple(dataIds)` ?

TallJimbo replied:
I thought about that a lot.  Since sometimes these will be pretty big collections, I preferred trusting the user to logically transfer ownership over making a (shallow) copy whether one was needed or not.  But maybe that's just premature optimization.

timj replied:
maybe update the docstring then to say that the caller should not modify it

timj commented on python/lsst/daf/butler/core/dimensions/coordinate.py:
```diff
@@ -19,83 +19,326 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+#
+# Design Notes for DataCoordinate
```
There's a lot of good information in here that might be better off in real documentation.

TallJimbo replied:
This reminded me that @jonathansick pointed me at https://jira.lsstcorp.org/browse/DM-21228, which recommends a `docs/dev` directory.  I'll move this there and leave a pointer to that as a code comment.

timj commented on python/lsst/daf/butler/core/dimensions/coordinate.py:
```diff
@@ -19,83 +19,326 @@
+# quite different from the other.  There are two big reasons for this:
+#
+# - We want this object to behave much like a `dict`, because plain `dict`
+#   objects are frequently used as informal and data IDs, but Python's
```
grammar: "used as informal and data IDs"

timj commented on python/lsst/daf/butler/core/dimensions/coordinate.py:
```diff
@@ -19,83 +19,326 @@
+#   the canonical set of identifiers, and hence we can't tell whether that data
+#   ID is equivalent to some other one until we've done a database query to get
+#   to canonical form.  I'm hoping that when that day comes we'll not try to
+#   make those kinds of data IDs into DataCoordinates, but even an adjacent
```
I was assuming that we'd do this by having special code that sees an instrument in a dataId, looks up the instrument class, runs a dataId demangler (that combines dayObs and seqNum into a exposure_id) and then returns a proper DataCoordinate. From that point on dayObs would not exist.

TallJimbo replied:
That's consistent with my thoughts as well.  If the demangling never needs a `Registry`, no problem.  If it does, we'll need to see if the layers that currently transform dictionaries into `DataCoordinates` already have a `Registry` to use for that, and if not, whether we should defer that transformation to a lower level or find a way to pass a `Registry` to them.

timj commented on python/lsst/daf/butler/core/dimensions/coordinate.py:
```diff
@@ -19,83 +19,326 @@
+# it fetch everything it needs, regardless of whether some of it was already
+# present, or is the complexity of trying to compute a diff between what it has
+# and what it needs worthwhile?  Should it fetch more than it needs, because
+# one fetch for many records can be a lot efficient than N separate fetches?
```
lot -> lot more

timj commented on python/lsst/daf/butler/core/dimensions/coordinate.py:
```diff
@@ -19,83 +19,326 @@
+# and what it needs worthwhile?  Should it fetch more than it needs, because
+# one fetch for many records can be a lot efficient than N separate fetches?
+#
+# The second version is what we've had in the codebase for over a year as of
```
Just say "since June 2019" rather than "over a year as of this writing"

timj commented on python/lsst/daf/butler/core/dimensions/coordinate.py:
```diff
@@ -111,115 +354,169 @@ def standardize(mapping: Optional[Union[Mapping[str, Any], NamedKeyMapping[Dimen
+            ``graph.implied`` is empty, and ``dataId.hasRecords()`` will never
+            return `True`.
+        """
+        assert len(graph.required) == len(values)
```
If this ever triggers people are going to be confused. Can we have an explicit RuntimeError with error message instead?

TallJimbo replied:
How about I add a message to the `assert`?  This is an internal API, and I feel like checking declared preconditions is one of the things `assert` is for - any code that calls this without guaranteeing it (doing the checking of its own arguments if necessary) is broken.

timj replied:
I guess I really dislike assert as a thing and I don't really understand what's wrong with:
```python
if len(graph.required) != len(values):
   raise AssertionError("Some message that tells me what len(a) was something and that this is a worry")
```
a message would always help though.

TallJimbo replied:
That's fair; I actually prefer `assert` because it raises `AssertionError` and that's the intent I want to convey, not because it can be disabled.  I think we've discussed this before: https://lsstc.slack.com/archives/C2JPL8U78/p1576168318008800, where raising `AssertionError` directly was considered (at least by one person) confusing, and I think the consensus was that `assert` was reasonable.  _I_ don't find raising `AssertionError` confusing, but I'm probably not the best judge of that for my own code.

Basically, I think `assert` is a really important _concept_, and I'm reluctant to avoid the language construct for it just because we're worried it _could_ be disabled, especially since I don't think being able to disable this check is obviously wrong.  It certainly could appear in a tight loop, and while I very much doubt the check is expensive, using `assert` means I don't have to _care_ if it's expensive.

TallJimbo replied:
I've added messages to all of the asserts in this module but am leaving these asserts for now.

timj commented on python/lsst/daf/butler/core/fileTemplates.py:
```diff
@@ -430,16 +430,18 @@ def format(self, ref: DatasetRef) -> str:
+        # This guards against Nones being explicitly present in the data ID
+        # (which can happen if, say, an exposure has no filter), as well as
+        # the case where only required dimensions are present (which in this
+        # context should only happen in unit tests; in generally we need all
```
generally -> general

timj commented on python/lsst/daf/butler/core/fileTemplates.py:
```diff
@@ -448,12 +450,12 @@ def format(self, ref: DatasetRef) -> str:
             if len(skypix) == 1:
-                fields["skypix"] = fields[skypix[0]]
-
-        # Extra information that can be included using . syntax
```
Can you attach this comment to where extras is now set?

timj commented on python/lsst/daf/butler/registry/_registry.py:
```diff
@@ -778,34 +779,53 @@ def expandDataId(self, dataId: Optional[DataId] = None, *, graph: Optional[Dimen
+                    )
+                    fetched = tuple(storage.fetch(dataIdSet))
+                    try:
+                        (record,) = fetched
```
This is a nice way to ensure only one item is returned.

timj commented on tests/test_dimensions.py:
```diff
@@ -212,5 +251,388 @@ def testPickling(self):
+        """
+        if dataIds is None:
+            dataIds = self.allDataIds
+        return DataCoordinateSequence(self.rng.sample(dataIds, n),
```
This reminded me that we have code in `python/lsst/daf/butler/tests/_testRepo.py` that uses numpy for this for no reason at all (it's the only explicit usage of numpy in the butler) where `rng.sample(range(10), 10)` would work fine.

timj reviewed (approved):
Looks like a nice cleanup.

TallJimbo committed:
Move DataCoordinate design notes to docs directory.

I've renamed the "Contributing" section to Design and Development
while adding a link to the new page in that section.  This reflects
the current recommendations in the DM Developer Guide.
